### ISSUE TITLE
[analyzer]  no analyzer error for `const` keyword, but dartpad fails with "Error: Constant evaluation error"

### ISSUE THREAD

**Issue by incendial** [dart-lang/sdk#55008]
Consider the following example:

```dart
extension type const Opt<T>._(({T value})? _) {
  factory Opt(T value) = Val<T>;
  factory Opt.none() = Non<T>;
}

extension type const Val<T>._(({T value}) _) implements Opt<T> {
  const Val(T value) : this._(const (value: value)); 

  T get value => _.value;
}
```

`const Val(T value) : this._(const (value: value)); ` gives no warning from the analyzer, but when code is run on dartpad it results in an error:

<img width="720" alt="Screenshot 2024-02-24 at 17 08 16" src="https://github.com/dart-lang/sdk/assets/21116742/1611b835-5ec2-4267-90f1-5a00bfd6c6f4">

<img width="523" alt="Screenshot 2024-02-24 at 17 08 28" src="https://github.com/dart-lang/sdk/assets/21116742/f35a52ad-dd0d-4ede-a003-46fb4e07964f">

Tested on dartpad main channel.

**Comment by lrhn** [comment]
This is a bug, IMO. 

The spec is a little too quiet on what `const` in front of a record, but it has to mean _something_!

You are allowed to have a `const` in front of a record expression, according to the [specification](https://github.com/dart-lang/language/blob/main/accepted/3.0/records/feature-specification.md#record-expressions). The spec then doesn't say what that *means*, it rather says that you don't need the `const` at all, because the expression will be constant if and only if the field expressions are. Still, if you're allowed to write it, and if there is any kind of consistency, it should put the expression into a constant context.

If the record expression here is in a `const` context, then the variable `value` cannot be referenced, since it's not a constant expression.

**Comment by bwilkerson** [comment]
@scheglov

**Comment by srawlins** [comment]
And cc @kallentu 😁

**Comment by eernstg** [comment]
Proposed change: https://github.com/dart-lang/language/pull/3630.

**Comment by eernstg** [comment]
Landed dart-lang/language#3630.
